### PR TITLE
User can sign up and log in

### PR DESCRIPTION
**Pull request by larathordar**
#### Description
Changes proposed in this pull request:

* Devise gem added
* log in and sign up feature added
#### What I have learned working on this feature:
To use Devise for user authentication in a rails app

#### Screenshots:
screen shot 2018-03-15 at 14 30 11
screen shot 2018-03-15 at 14 30 26

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
```
Comment (holgertidemand) [inline app/models/user.rb]: Remove comments

File: features/step_definitions/basic_steps.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@
   end
 end
 
-Given("I am on the {string} page") do |page_name|
+Given("I am on the {string} Page") do |page_name|
```
Comment (holgertidemand) [inline features/step_definitions/basic_steps.rb]: change back to lowercase ```page ```

File: app/views/index/index.html.erb [hunk 3]
```diff
@@ -1 +1,15 @@
...
+  <% else %>
+    <li>
+      <%= link_to 'Login', new_user_session_path %>
+    </li>
+    <li>
+      <%= link_to 'Register', new_user_registration_path %>
+    </li>
+  <% end %>
+
+<h1>Dream Team Publications</h1>
```
Comment (holgertidemand) [inline app/views/index/index.html.erb]: Remove h1. you have two

File: features/step_definitions/basic_steps.rb [hunk 4]
```diff
@@ -31,10 +31,12 @@
 def page_path(path)
   if path == 'Create Article'
     new_article_path
+  elsif path == 'Subscriber'
+      new_user_registration_path
   elsif path == 'Holger is the best'
     article_title = Article.find_by(title: path)
     article_path(article_title)
-  elsif path == 'Index'
+  elsif path == 'index'
```
Comment (holgertidemand) [inline features/step_definitions/basic_steps.rb]: Change to UpperCase ``` Index``` Here and in the test. We used that.

File: features/step_definitions/basic_steps.rb (hunk 2 above)
Comment (holgertidemand) [inline features/step_definitions/basic_steps.rb]: ``` page ``` change to lowercase p

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   root controller: :index, action: :index
   resources :articles, only: [:new, :create, :show, :destroy]
+  devise_for :users
+  root controller: :index, action: :index
```
Comment (jyuw) [inline config/routes.rb]: This line already exists on line 2

File: features/step_definitions/login_steps.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+Given("following user exist") do |table|
+  table.hashes.each do |user|
+    FactoryBot.create(:user, user)
```
Comment (holgertidemand) [inline features/step_definitions/login_steps.rb]: Make use of short command for Factorybot ```  create(:user, user) ```

File: features/user_can_register.feature [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+          And I fill in "Password" with "hellooo"
+          And I click on "Sign up"
+          Then I should see "1 error prohibited this user from being saved: Password confirmation doesn't match Password"
+
+        Scenario: Visitor fills in invalid password
+          When I fill in "Email" with "john@smith.com"
+          And I fill in "Password" with "pas"
+          And I fill in "Password confirmation" with "pas"
+          And I click on "Sign up"
+          Then I should see "1 error prohibited this user from being saved: Password is too short"
```
Comment (holgertidemand) [inline features/user_can_register.feature]: Format this file. Make it beautiful

File: features/step_definitions/login_steps.rb (hunk 6 above)
Comment (jyuw) [reply features/step_definitions/login_steps.rb]: This step definition already exists in register_steps!

File: features/user_can_login.feature [hunk 8]
```diff
@@ -0,0 +1,33 @@
+Feature: User login
+  As a user
+  In order to access my account
+  I want to be able to login
+
+  Background:
+    Given following user exist
+    | email            | password |
```
Comment (jyuw) [inline features/user_can_login.feature]: Make sure the pipes are lining up and pretty (:

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (holgertidemand) [inline spec/models/user_spec.rb]: Remove this file

Comment (magnus-thor) [reply spec/models/user_spec.rb]: You need this file to unit test the user model. And you should have tests in there now.

File: spec/models/model_spec.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Model, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (magnus-thor) [inline spec/models/model_spec.rb]: Do you have a Model called model? If not you can delete this file

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -1,6 +1,8 @@
 <!DOCTYPE html>
 <html>
   <head>
+    <p class="notice"><%= notice %></p>
+    <p class="alert"><%= alert %></p>
```
Comment (magnus-thor) [inline app/views/layouts/application.html.erb]: This should not be inside the head tags

File: config/routes.rb [hunk 12]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  root controller: :index, action: :index
```
Comment (magnus-thor) [inline config/routes.rb]: You already have this.

File: app/views/index/index.html.erb [hunk 13]
```diff
@@ -1 +1,13 @@
...
+<% if user_signed_in? %>
+    <li><%= link_to 'Logout', destroy_user_session_path, method: :delete %></li>
+  <% else %>
+    <li>
+      <%= link_to 'Login', new_user_session_path %>
+    </li>
+    <li>
+      <%= link_to 'Register', new_user_registration_path %>
+    </li>
+  <% end %>
```
Comment (tochman) [inline app/views/index/index.html.erb]: Formatting seems off...

File: features/user_can_register.feature [hunk 14]
```diff
@@ -0,0 +1,41 @@
+Feature: User can register
+As a user
+In order to access my account
+I would like to have a signup/login function
```
Comment (tochman) [inline features/user_can_register.feature]: Login flow is already tested in another feature file. What does your PT story tell you and why?

File: features/user_can_register.feature [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    And I click on "Sign up"
+    Then I should see "2 errors prohibited this user from being saved: Password can't be blankPassword confirmation doesn't match Password"
+
+  Scenario: Visitor fills in all fields except Password confirmation [Sad Path]
+    Given I fill in "Email" with "john@smith.com"
+    When I fill in "Password" with "hellooo"
+    And I click on "Sign up"
+    Then I should see "1 error prohibited this user from being saved: Password confirmation doesn't match Password"
+
+  Scenario: Visitor fills in invalid password
```
Comment (tochman) [inline features/user_can_register.feature]: You add the words `[Sad Path]` to every scenario but the first one in this file. On this one, you don't. Why? 
Also, please take a look on the other feature file in this PR, where you also omit this. Please choose a standard and stick to it.

File: app/views/index/index.html.erb (hunk 13 above)
Comment (Fleurtam) [reply app/views/index/index.html.erb]: Fixed the indentation

File: features/user_can_register.feature (hunk 15 above)
Comment (Fleurtam) [reply features/user_can_register.feature]: Fixed consistency

File: features/user_can_register.feature (hunk 14 above)
Comment (Fleurtam) [reply features/user_can_register.feature]: Removed login from the feature

File: features/step_definitions/register_steps.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+Then("I should be redirected to {string} page") do |page_name|
+  expect(page.current_path).to eq page_path(page_name)
+end
```
Comment (holgertidemand) [inline features/step_definitions/register_steps.rb]: We have a step in basic_steps that do the same thing. 
```
Then("I should be on the {string} page") do |article_title|
  expect(page.current_path).to eq find_article(article_title)
end
``` 

Please refactor test and remove this

File: app/views/index/index.html.erb [hunk 17]
```diff
@@ -1 +1,13 @@
...
+<% if user_signed_in? %>
+    <li><%= link_to 'Logout', destroy_user_session_path, method: :delete %></li>
+<% else %>
+    <li>
+      <%= link_to 'Login', new_user_session_path %>
+    </li>
+    <li>
+      <%= link_to 'Register', new_user_registration_path %>
+    </li>
+<% end %>
```
Comment (magnus-thor) [inline app/views/index/index.html.erb]: Be consistent with the Indentation, some lines have 4 spaces other 2 spaces

File: features/user_can_register.feature [hunk 18]
```diff
@@ -0,0 +1,41 @@
+Feature: User can register
+As a user
+In order to access my account
+I would like to have a signup function
```
Comment (magnus-thor) [inline features/user_can_register.feature]: Fix the Indentation

**Review by holgertidemand — APPROVED** [review]

**Review by magnus-thor — COMMENTED** [review]
Looks Good.

**Review by tochman — APPROVED** [review]

**Pull request merged**